Support graceful exit if dirs missing and --allow-no-dir

OliverUv opened the pull request:
Useful for my npm-link in docker workflow. Will probably do a writeup about it later on.

This commit introduces the first flag with multiple words in it, chose to go with choo-choo-train case, since the `--npm-link-option`s are written in that case.

I suggest refactoring the api function to use dictionary options later next time you break backwards compatibility. Something like

```js
function api(sharedDir, targetDir, options) {
  var opts = {
    'moduleList': [],
    'linkOptions': [], // previously "optionList"
    'executable': 'npm',
    'allowNoDir': false,
  };
  Object.keys(options).forEach(function extend_opts(key) {
    opts[key] = options[key];
  });

  // rest of function uses opts to access options
```

I guess this could be done without breaking backwards compat, but it'd be ugly.

OliverUv committed:
Support graceful exit if dirs missing and --allow-no-dir

OrKoN commented:
@OliverUv Hey, do you think there are scenarios where linking is possible w/o shared & target dirs existing? Perhaps it can be improved w/o an extra flag?

OliverUv commented:
I'm not sure I understand what you mean.

If I implemented this without regard for backwards compatibility, I'd just skip the flag and always exit cleanly if dirs are missing (similar to what happens now if sharedDir is empty). Because I'm thinking the whole idea of this module is dependent on having a shared and target dir. I can implement this without using the flag if you feel like breaking backwards compatibility is ok. (Since we print a notice about the directory not existing, any misconfiguration of a CI environment should be obvious from looking at build logs)

OrKoN commented:
I think 0.5.0 is exactly meant for breaking changes. Feel free to go ahead
and add the feature w/o the flag.
Oliver Uvman <notifications@github.com> schrieb am Mo. 27. März 2017 um
20:34:

> I'm not sure I understand what you mean.
>
> If I implemented this without regard for backwards compatibility, I'd just
> skip the flag and always exit cleanly if dirs are missing (similar to what
> happens now if sharedDir is empty). Because I'm thinking the whole idea of
> this module is dependent on having a shared and target dir. I can implement
> this without using the flag if you feel like breaking backwards
> compatibility is ok. (Since we print a notice about the directory not
> existing, any misconfiguration of a CI environment should be obvious from
> looking at build logs)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OrKoN/npm-link-shared/pull/12#issuecomment-289543858>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYXLoL4qL56fm937UbwVxl_4xy6Im6Wks5rqAEdgaJpZM4MqAaQ>
> .
>

OliverUv commented:
Depends on what we consider a breaking change. We'll switch away from exiting with an error code to printing an error and exiting cleanly. If we consider this a breaking change (and I guess I would, considering that this might be in use in all sorts of unpredictable CI systems) then we should bump this to 1.0.0 (according to semver). I'd be happy to do this.

OliverUv commented:
(according to semver, since our major version is `0`, anything is allowed to break at any time, but I think most people assume that breaking changes will only occur when major version changes).

I'll send a pull request with the changes I have in mind, sans version changes, and let you take that decision.

OrKoN commented:
afaik according to semver if it is less than 1.0.0, every minor version is
allowed to contain breaking changes. And npm will not install 0.5.0 for
^0.4.0
Oliver Uvman <notifications@github.com> schrieb am Di. 28. März 2017 um
16:53:

> (according to semver, since our major version is 0, anything is allowed
> to break at any time, but I think most people assume that breaking changes
> will only occur when major version changes).
>
> I'll send a pull request with the changes I have in mind, sans version
> changes, and let you take that decision.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OrKoN/npm-link-shared/pull/12#issuecomment-289795985>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYXLm9iY6uZvqWIf2yeixvBuCoL_dnAks5rqR7QgaJpZM4MqAaQ>
> .
>

OliverUv committed:
Remove --allow-no-dir option, change warning output color

OliverUv committed:
Refactor api function arguments

OliverUv committed:
Do not change Array prototype

Altering system prototypes is usually frowned upon. It can stop V8 from
optimizing, and can surprise other developers.

OliverUv committed:
Update README

OliverUv commented:
New changes do what the commit messages say. You can see what the new README looks like rendered here: https://github.com/OliverUv/npm-link-shared

All in all, I think these are good changes. I also manually tested the cli changes by commenting out all the `link` calls in the test files, running the equivalent things from the command line, and observing that the tests succeeded.

One thing to note: The usage output previously said that `npm-link-shared shared_dir target_dir --production module-a` would be ok input. This would not work, as it would be translated by minimist to `--production=module-a`. I have changed the documents to specify that any `--` flags must appear **after** the shared dir, target dir, and modules.

OrKoN commented:
Thanks, published as 0.5.0
